executor/aggfuncs/func_max_min.go

```diff
@@ -14,49 +14,126 @@
 package aggfuncs
 
 import (
+	"container/heap"
+
 	"github.com/pingcap/tidb/sessionctx"
 	"github.com/pingcap/tidb/types"
 	"github.com/pingcap/tidb/types/json"
 	"github.com/pingcap/tidb/util/chunk"
 	"github.com/pingcap/tidb/util/stringutil"
 )
 
+type maxMinHeap struct {
+	data    []interface{}
+	h       heap.Interface
+	varSet  map[interface{}]int64
+	isMax   bool
+	cmpFunc func(i, j interface{}) int
+}
+
+func newMaxMinHeap(isMax bool, cmpFunc func(i, j interface{}) int) *maxMinHeap {
+	h := &maxMinHeap{
+		data:    make([]interface{}, 0),
+		varSet:  make(map[interface{}]int64),
+		isMax:   isMax,
+		cmpFunc: cmpFunc,
+	}
+	return h
+}
+
+func (h *maxMinHeap) Len() int { return len(h.data) }
+func (h *maxMinHeap) Less(i, j int) bool {
+	if h.isMax {
+		return h.cmpFunc(h.data[i], h.data[j]) > 0
+	}
+	return h.cmpFunc(h.data[i], h.data[j]) < 0
+}
+func (h *maxMinHeap) Swap(i, j int) { h.data[i], h.data[j] = h.data[j], h.data[i] }
+
+func (h *maxMinHeap) Push(x interface{}) {
+	h.data = append(h.data, x)
+}
+func (h *maxMinHeap) Pop() interface{} {
+	old := h.data
+	n := len(old)
+	x := old[n-1]
+	h.data = old[0 : n-1]
+	return x
+}
+
+func (h *maxMinHeap) Top() (val interface{}, isEmpty bool) {
+retry:
+	if h.Len() == 0 {
+		return nil, true
+	}
+	top := h.data[0]
+	if h.varSet[top] == 0 {
+		_ = heap.Pop(h)
+		goto retry
+	}
+	return top, false
+}
+
+func (h *maxMinHeap) Append(val interface{}) {
+	h.varSet[val]++
+	if h.varSet[val] == 1 {
+		heap.Push(h, val)
+	}
+}
+func (h *maxMinHeap) Remove(val interface{}) {
+	if h.varSet[val] > 0 {
+		h.varSet[val]--
+	}
+}
+func (h *maxMinHeap) Reset() {
+	h.data = h.data[:0]
+	h.varSet = make(map[interface{}]int64)
+}
+
 type partialResult4MaxMinInt struct {
 	val int64
 	// isNull is used to indicates:
 	// 1. whether the partial result is the initialization value which should not be compared during evaluation;
 	// 2. whether all the values of arg are all null, if so, we should return null as the default value for MAX/MIN.
 	isNull bool
+	// maxMinHeap is a ordered queue, using to evaluate the maximum or minimum value in a sliding window.
+	*maxMinHeap
```

Comment:
Suggest keeping this entry explicitly.
```suggestion
	heap *maxMinHeap
```